src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1587,16 +1587,40 @@ public DynamicObject replace(DynamicObject array, DynamicObject other,
 
         @Specialization(guards = "strategy.matches(array)", limit = "ARRAY_STRATEGIES")
         public DynamicObject rotate(DynamicObject array, int rotation,
-                @Cached("of(array)") ArrayStrategy strategy) {
-            final int size = strategy.getSize(array);
+                @Cached("of(array)") ArrayStrategy strategy,
+                @Cached("createIdentityProfile()") IntValueProfile sizeProfile,
+                @Cached("createIdentityProfile()") IntValueProfile rotationProfile) {
+            final int size = sizeProfile.profile(strategy.getSize(array));
+            rotation = rotationProfile.profile(rotation);
             assert 0 < rotation && rotation < size;
             final ArrayMirror mirror = strategy.newMirror(array);
 
-            rotateReverse(rotation, size, mirror);
+            if (CompilerDirectives.isPartialEvaluationConstant(size) &&
+                    CompilerDirectives.isPartialEvaluationConstant(rotation) &&
```

Comment:
Interesting. I didn't know about these. I probably would've just used two different `@Specialization`s, but this cuts down on duplicating the boilerplate.